### PR TITLE
[oneMKL][spblas] Add `optimize_symv()` and `optimize_gemm()` routines in oneMKL specification.

### PR DESCRIPTION
**Pull request by baeseung-intel**
**Goal:**
In this PR, I'd like to add two optimization routines, `optimize_symv()` and `optimize_gemm()` into oneMKL specification.

**Details:**

- Add `optimize_symv()` API into oneMKL specification, which prepares optimization for `oneapi::mkl::sparse::symv()` API.
- Add `optimize_gemm()` API into oneMKL specification, which prepares optimization for `oneapi::mkl::sparse::gemm()` API.
- For `optimize_gemm()` API, we propose two different routines; the first API will prepare optimization based on sparse input matrix only, and the other API will prepare optimization based on both sparse and dense input matrices.
- Add those two APIs in the `Sparse BLAS` main page.

### CODE REVIEW

File: source/elements/oneMKL/source/domains/spblas/gemmoptimize.rst [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+=============
+
+Performs internal optimizations for oneapi::mkl::sparse::gemm by analyzing
+the matrix structure.
+
+.. rubric:: Description and Assumptions
+
+The oneapi::mkl::sparse::optimize_gemm routine analyzes matrix structure
+and performs optimizations. Optimized data is then stored in
+the handle.
```
Comment (spencerpatty) [inline source/elements/oneMKL/source/domains/spblas/gemmoptimize.rst:17]: lets explicitly add some language that discusses the two options here, and why either one exists (as well as talking about how the AB one should support calling with multiple B input size/shapes ...)

Comment (baeseung-intel) [reply source/elements/oneMKL/source/domains/spblas/gemmoptimize.rst:17]: Added a paragraph below.

**Comment by rscohn2** [comment]
I would like to wrap up the 1.3-provisional release. Is this ready for merging.

**Comment by spencerpatty** [comment]
@rscohn2 let me take one more pass at cleaning up language a little bit,  it is 95% there in my opinion... :)

File: source/elements/oneMKL/source/domains/spblas/gemmoptimize.rst [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+which are done only based on Sparse matrix structure,
+two version of ``optimize_gemm`` routines are listed here for preparing optimizations
+for oneapi::mkl::sparse::gemm routine,
+since we might perform optimizations for oneapi::mkl::sparse::gemm
+with respect to the shape and layout of the given dense matrix
+as well as based on sparse matrix structure only.
+If user wants to run multiple times oneapi::mkl::sparse::gemm with the same shape and layout
+of input dense matrix :math:`B`, then the latter version of ``optimize_gemm`` routine might
+give further performance optimization specific to the input dense matrix as well as
+the given sparse matrix.
```
Comment (spencerpatty) [inline source/elements/oneMKL/source/domains/spblas/gemmoptimize.rst:29]:
```suggestion
In contrast to other optimization routines in Sparse BLAS domain,
which are done solely based on the sparse matrix pattern, 
two versions of ``sparse::optimize_gemm`` routines are provided for preparing different
 optimizations for ``sparse::gemm`` routine.  In particular, if the shape
 of the dense matrix right hand side, :math:`B`, is unknown or widely varying in 
 subsequent calls to ``sparse::gemm`` then a user might reasonably
only wish to perform optimizations for ``:sparse::gemm`` with respect 
to the sparse matrix structure, but if one or more particular shapes of :math:`B` 
is available, then each :math:`B` shape can be provided as an additional hint 
along with the sparse matrix pattern in the call to ``sparse::optimize_gemm``.  This 
second version of the API with :math:`B` shape should be callable one or 
more times and may allow libraries to provide more targeted performance 
optimizations.
```

File: source/elements/oneMKL/source/domains/spblas/gemmoptimize.rst [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+Performs internal optimizations for oneapi::mkl::sparse::gemm by analyzing
+the matrix structure.
+
+.. rubric:: Description and Assumptions
+
+The oneapi::mkl::sparse::optimize_gemm routine analyzes matrix structure
+and performs optimizations. Optimized data is then stored in
+the handle.
+
+In contrast to other optimization routines in Sparse BLAS domain,
```
Comment (spencerpatty) [inline source/elements/oneMKL/source/domains/spblas/gemmoptimize.rst:19]:
```suggestion
In contrast to other optimization routines in Sparse BLAS domain
```

File: source/elements/oneMKL/source/domains/spblas/gemmoptimize.rst [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+which are done solely based on the sparse matrix pattern, 
+two versions of ``sparse::optimize_gemm`` routines are provided for preparing different
```
Comment (spencerpatty) [inline source/elements/oneMKL/source/domains/spblas/gemmoptimize.rst:21]:
```suggestion
two versions of the ``sparse::optimize_gemm`` routine are provided for preparing different
```

File: source/elements/oneMKL/source/domains/spblas/gemmoptimize.rst [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+ optimizations for ``sparse::gemm`` routine.  In particular, if the shape
+ of the dense matrix right hand side, :math:`B`, is unknown or widely varying in 
+ subsequent calls to ``sparse::gemm`` then a user might reasonably
+only wish to perform optimizations for ``:sparse::gemm`` with respect 
+to the sparse matrix structure, but if one or more particular shapes of :math:`B`
```
Comment (spencerpatty) [inline source/elements/oneMKL/source/domains/spblas/gemmoptimize.rst:26]:
```suggestion
to the sparse matrix structure.  However, if one or more particular shapes of :math:`B` 
```

**Review by spencerpatty — APPROVED** [review]
I approve of this addition to the Sparse BLAS spec :)

**Pull request merged**